---
title: "From zero to bioinformatics in R interactively"
author: "Jonathan Dreyfuss PhD, Dir. Bioinfo. & Biostat. Core, Joslin Diabetes Center"
output: learnr::tutorial
runtime: shiny_prerendered
tutorial:
  version: 0.0.2.1009
---

```{r setup, include=FALSE}
library(BiocManager)
library(limma)
library(ezlimma)
library(ezlimmaplot)
library(ggplot2)
library(learnr)
options(repos = BiocManager::repositories(), stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = FALSE)
```

## Introduction

### Welcome

You'd like to analyze and visualize your data. Your workflow will be the most flexible, powerful, and reproducible if it involves *programming*. This type of programming is now often called *data science*.

### Goal of tutorial
Programming looks intimidating, but once you understand the basics, you should be able to Google questions, understand the answers, and be productive. I'll try to teach as much of the basics as I can in the hour I have to present this tutorial. You may wish to come back to it later to refresh and practice.

## R & RStudio
### R
- R is a free language and software for statistical computing and graphics  
- R has thousands of free contributed packages, such as roughly 2,000 bioinformatics packages that make up the [Bioconductor](https://www.bioconductor.org/) project, and packages that allow R to create data-rich, interactive [websites](https://shiny.rstudio.com/gallery/)  
- You can download base R for [Windows](https://cloud.r-project.org/bin/windows/base/) or [Mac OS X](https://cloud.r-project.org/bin/macosx/) or Linux

### RStudio
- [RStudio](https://www.rstudio.com/) is the most popular integrated development environment (IDE) for R  
- This document was created by using R through RStudio  
- You can download [RStudio Desktop: Open Source Licence](https://www.rstudio.com/products/rstudio/download/) for free
- RStudio has a lot of functionality ![RStudio](./images/rstudio-workbench.png)

### RStudio panels
R code is in the *Source* panel, which is at the top left. This is where you write & save your code. The *Console* is below; it executes R code. Press ctrl+enter (or cmd+enter on a mac) to send a line or highlighted portion of code from the editor to the console. When a line is executed, its return value is displayed, unless that value is assigned to a variable.

```{r one-plus-one, exercise=TRUE, exercise.eval = TRUE}
1+1
```

```{r comment, exercise=TRUE, exercise.eval = TRUE}
# is a comment. it doesn't get executed.
```

```{r one-plus-onex, exercise=TRUE, exercise.eval = TRUE}
x <- 1+1 #nothing is displayed
```

### R Markdown
This HTML is written in [R Markdown](https://rmarkdown.rstudio.com/), which weaves together text, figures, and code. You can "Knit" R Markdown documents into HTML, PDF, Microsoft Word, or other formats by pressing the "Knit" icon in R Studio once you have installed all necessary packages. You can install a package, such as the `knitr` package, with `install.packages("knitr", repos = "https://cloud.r-project.org")`.

![Image from [RStudio R Markdown lesson](https://rmarkdown.rstudio.com/lesson-9.html)](./images/knit.png)

### Resources
In RStudio go to Help/R Help, then

- [Learning R Online](https://www.rstudio.com/online-learning/). One resource they cite is the free book [R for Data Science](https://r4ds.had.co.nz/), which has a particularly helpful chapter on [data visualization](https://r4ds.had.co.nz/data-visualisation.html) using the popular R package `ggplot2`.  
- *An Introduction to R*. The intro is also available in [PDF](http://cran.r-project.org/doc/manuals/R-intro.pdf). I found it (especially the first six chapters, which is only 30 pages) very helpful when I was learning R.  
- RStudio also maintains many [cheatsheets](https://www.rstudio.com/resources/cheatsheets/), including an introductory cheatshet for [Base R](./base-r.pdf) -- this cheatsheet has much of what we will cover.  
- Chan HSPH R/bioinformatics holds regular [trainings](http://bioinformatics.sph.harvard.edu/training/)

## R packages developed by the Core
Dr. Hui Pan (Senior Bioinformatician, Bioinfo. & Biostat. Core) and I have developed several R packages to streamline bioinformatics analysis. These are based on the popular R package [limma](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4402510/), which applies linear regression modeling of omics data with sophisticated variance estimation to improve power, and has been validated in multiple data types. Our packages are freely available on the popular code repository [Github](https://github.com/). R packages on Github can be directly installed in R.

Our analysis package is [ezlimma](https://github.com/jdreyf/ezlimma) and our plotting package is [ezlimmaplot](https://github.com/jdreyf/ezlimmaplot), whose README includes installation instructions for both.

## Types
### Numeric 
```{r Numeric, exercise=TRUE, exercise.eval = TRUE}
n <- 4
print(n) # displays the variable
```

This assigns the value 4 to variable `x`. `<-` represents an arrow.
```{r}
n <- 4 #Create a global variable n=4 to be used later on.
```

You can modify the code below to, say, `n=5` and you can see the output change when you press on the button `Run Code`. This modification will only affect the output of this exercise and no others, by design of the website. However, if you change a variable's value in R, the value will remain changed.

```{r Numeric1, exercise=TRUE, exercise.eval = TRUE}
n = 4 
n #same as print() when not in loop or function
```

`=` does the same thing.  `<-` is usually equivalent to `=`.

```{r Numeric4, exercise=TRUE, exercise.eval = TRUE}
n+2
```

Alternative expression of above:

```{r Numeric5, exercise=TRUE, exercise.eval = TRUE}
sum(n, 2)
```

However, R is case sensitive, so this produces an error
```{r Numeric6, eval=FALSE, exercise=TRUE}
N+2 #Error: object 'N' not found
```

### Logicals
A logical can have the values `TRUE`, `FALSE`, and `NA` ("Not Available"). You might see `TRUE` and `FALSE` abbreviated as `T` and `F`, respectively, but I don't recommend using the abbreviation. You can create logicals with: `<`, `<=`, `>`, `>=`, `==` for equality, `!=` for inequality. Logical operators are: `!` for *not*; `&` for *and*; `|` for *or*.
```{r logic, exercise=TRUE, exercise.eval = TRUE}
4==4 #is 4 equal to 4?
```

```{r logic01, exercise=TRUE, exercise.eval = TRUE}
4!=5 #is 4 not equal to 5?
```

Assigns the value of `4>5` (i.e. `FALSE`) to `b`.
```{r logic02, exercise=TRUE, exercise.eval = TRUE}
b <- 4 > 5
b
``` 

```{r}
b <- 4 > 5 #Create a global variable n1=4 > 5 (i.e. FALSE)
``` 

```{r logic03, exercise=TRUE, exercise.eval = TRUE}
!b #not b
```

```{r logic04, exercise=TRUE, exercise.eval = TRUE}
b | (5>4)
```

```{r logic05, exercise=TRUE, exercise.eval = TRUE}
b & TRUE
```

An easy mistake is to check equality with `=`.
```{r logic2, exercise=TRUE}
4=4
```

This doesn't work, `=` is for assigning values to variables. Instead, use `==`.

We use logicals in `if` statements.

```{r logic4, exercise=TRUE, exercise.eval = TRUE}
z <- 3
if (z==3){ 
  print('z is 3') 
}
```

```{r}
z <- 3 #Create a global variable z=3
```

We can also add an action if our condition is `FALSE`.

```{r logic5, exercise=TRUE, exercise.eval = TRUE}
if (z==4){ 
  print('z is 4') 
} else {
  print('z is not 4')
}
```

### Characters
```{r char, exercise=TRUE, exercise.eval = TRUE}
x <- 'hello'
x
```
`'hello'` is a character string.

Get same behavior as above; you can use either single or double quotes
```{r char1, exercise=TRUE, exercise.eval = TRUE}
x <- "hello"
x
```

```{r}
x <- 'hello' #Create a global variable c='hello'
```

Can paste character strings together
```{r char2, exercise=TRUE, exercise.eval = TRUE}
paste(x, 'world')
```

### `NA` (Not Available) values
This is for missing data. `NA`s can occur in any variable type and are treated specially.
```{r NAs, exercise=TRUE, exercise.eval = TRUE}
u <- NA #data for u is missing
```

```{r}
u <- NA #global: data for u is missing
```

We don't know `u`, so the result is also missing.
```{r NAs1, exercise=TRUE, exercise.eval = TRUE}
u+1
```

We test if a value is `NA` using `is.na()`:

```{r NAs2, exercise=TRUE, exercise.eval = TRUE}
is.na(u)
```

Many functions can ignore `NA`s. For example, we can ignore `NA`s in `sum`:
```{r NAs3, exercise=TRUE, exercise.eval = TRUE}
sum(u, 1, na.rm=TRUE)
```

## Getting help
To get help on `sum()` function, can Google: "r sum" or "sum in r". If you know the function name, you can enter `?sum` in the console, which will open the help page for `sum` in R Studio's *Help* tab in the bottom right panel.

## Change types, called *coercion*
Functions sometimes need input to be of a particular type, for example a function may accept a character. If you instead have a number, you can coerce numbers to character.
```{r change, exercise=TRUE, exercise.eval = TRUE}
as.character(z)
```
We can tell the output is a character because it's now in quotes.

You can also change logical to numeric
```{r change1, exercise=TRUE, exercise.eval = TRUE}
as.numeric(TRUE) #1
```
So, the numeric value of `TRUE` is 1,

```{r change2, exercise=TRUE, exercise.eval = TRUE}
as.numeric(FALSE) #0
```

Whereas the numeric value of `FALSE` is 0.

## Vectors
We want a variable to be able to hold more than a single value. A vector is a list of values of a common type (e.g. numbers). We can do similar operations as above to multiple values in a vector simultaneously.

### Numeric vectors

We can create vectors with `c()`, which concatenates values.
```{r Vectors, exercise=TRUE, exercise.eval = TRUE}
x <- c(1, 2, 3, 4)
x
```

```{r}
x <- 1:4 #Create global variables to use later on
y <- 2*x+1 
b1 <- y > 6 
b2 <- y == 7
```

We can also use `:` to create a continuous sequence
```{r Vector01, exercise=TRUE, exercise.eval = TRUE}
x <- 1:4
x
```

### Practice using functions
A more general way than `:` to create sequences is `seq()` function. We can learn about this function with 

```{r seq, exercise=TRUE}
?seq # works in RStudio, but does not work well on shinyapps.io
```

The following calls are the same:
```{r Vector2, exercise=TRUE, exercise.eval = TRUE}
seq(1,4,1) #R expects arguments in the given order
```

```{r Vector21, exercise=TRUE, exercise.eval = TRUE}
seq(to=4, from=1, by=1) #But you can name them & give in any order
```

You don't always need to write out full argument names
```{r Vector22, exercise=TRUE, exercise.eval = TRUE}
#R matches "fr" to argument "from", since no other arguments start w/ "fr"
seq(to=4, fr=1, by=1)
```

You don't need to specify arguments that already have the proper default
```{r Vector23, exercise=TRUE, exercise.eval = TRUE}
seq(to=4, by=1) #since default has from=1
```

### Operations on numeric vectors
```{r Vector3, exercise=TRUE, exercise.eval = TRUE}
y <- 2*x+1 #operates on each element of x
y
```

```{r Vector32, exercise=TRUE, exercise.eval = TRUE}
mean(x)
```

```{r Vector33, exercise=TRUE, exercise.eval = TRUE}
length(x) #length of vector
```

```{r Vector34, exercise=TRUE, exercise.eval = TRUE}
var(x) #variance
```

```{r Vector35, exercise=TRUE, exercise.eval = TRUE}
t.test(x, y)
```

We can repeat values using `rep()`
```{r Vector36, exercise=TRUE, exercise.eval = TRUE}
rep(x, times=2)
```

```{r Vector37, exercise=TRUE, exercise.eval = TRUE}
rep(x, each=2)
```

### Logical vectors
```{r Vector4, exercise=TRUE, exercise.eval = TRUE}
b1 <- y > 6 #is each element of y greater than 6?
b1
```

```{r Vector40, exercise=TRUE, exercise.eval = TRUE}
b2 <- y == 7 #is each element of y equal to 7?
b2
```

```{r Vector41, exercise=TRUE, exercise.eval = TRUE}
b1 & b2 #and
```

```{r Vector42, exercise=TRUE, exercise.eval = TRUE}
b1 | b2 #or
```

```{r Vector43, exercise=TRUE, exercise.eval = TRUE}
!b1 #not b1
```

`which()` identifies the indices of the values that are `TRUE`
```{r Vector44, exercise=TRUE, exercise.eval = TRUE}
which(b1)
```

Some functions automatically coerce logicals to numeric
```{r Vector45, exercise=TRUE, exercise.eval = TRUE}
sum(b1)
```

```{r Vector46, exercise=TRUE, exercise.eval = TRUE}
mean(b1)
```

### Character vectors
```{r Vector5, exercise=TRUE, exercise.eval = TRUE}
first.names <- c('Frederick', 'Charles')
```

```{r Vector50, exercise=TRUE, exercise.eval = TRUE}
last.names <- c('Banting', 'Best')
```

```{r vector50-global}
first.names <- c('Frederick', 'Charles')
last.names <- c('Banting', 'Best')
```

`paste()` combines first & last names to make full names
```{r Vector51, exercise=TRUE, exercise.eval = TRUE}
paste(first.names, last.names)
```

### Names
Elements of vectors can have names. We can create a vector with names using the concatenation function `c()`.
```{r Vector6, exercise=TRUE, exercise.eval = TRUE}
v <- c(a=4, b=5, c=6)
v
```

```{r vectors6-global}
v <- c(a=4, b=5, c=6) #Global variable
```

```{r Vector61, exercise=TRUE, exercise.eval = TRUE}
names(v) #this is itself a vector
```

### Subset vectors
We can subset with square brackets using indices or names. Two ways to select the 1st element, 
```{r Vector7, exercise=TRUE, exercise.eval = TRUE}
v[1]
```

```{r Vector70, exercise=TRUE, exercise.eval = TRUE}
v["a"]
```

We can select 2nd & 3rd element in multiple ways
```{r Vector71, exercise=TRUE, exercise.eval = TRUE}
v[2:3]
```

```{r Vector72, exercise=TRUE, exercise.eval = TRUE}
v[c(2,3)]
```

```{r Vector73, exercise=TRUE, exercise.eval = TRUE}
v[c("b", "c")]
```

```{r Vector74, exercise=TRUE, exercise.eval = TRUE}
v[c(FALSE, TRUE, TRUE)]
```

Using `-` excludes indices. For example, `-1` returns all elements except the 1st.
```{r Vector75, exercise=TRUE, exercise.eval = TRUE}
v[-1]
```

## Matrices & Data frames
### Matrices
Matrix is a table of elements of a common type.
```{r Mat, exercise=TRUE, exercise.eval = TRUE}
m <- matrix(1:4, ncol=2, nrow=2) #creates a numeric matrix
m
```

```{r Mat02-setup}
m <- matrix(1:4, ncol=2, nrow=2) #Global variable for a numeric matrix m
```

```{r Mat02, exercise=TRUE, exercise.eval = TRUE}
2*m+1 #operates on each element
```

```{r Mat03, exercise=TRUE, exercise.eval = TRUE}
m <- matrix(c('a', 'b', 'c', 'd'), ncol=2, nrow=2) #creates a character matrix
m
```

We can also create matrices by column-binding vectors using `cbind()` or row-binding vectors using `rbind()`.
```{r Mat04, exercise=TRUE, exercise.eval = TRUE}
m <- cbind(first.names, last.names)
m
```

```{r}
m <- cbind(first.names, last.names) ##Global variable for a character matrix m with names
rownames(m) <- c("FB", "CB") 
```

Matrices can have `rownames` & `colnames`
```{r Mat2, exercise=TRUE, exercise.eval = TRUE}
colnames(m)
```

```{r Mat22, exercise=TRUE, exercise.eval = TRUE}
rownames(m) <- c("FB", "CB") #assigning rownames to m 
m
```

We can subset matrices & data frames using `object[rows, columns]`. As with vectors, we can specify `rows` or `columns` using indices or names.
```{r Mat3, exercise=TRUE, exercise.eval = TRUE}
m[1,] #1st row; leaving the columns index empty selects all columns
```

```{r Mat30, exercise=TRUE, exercise.eval = TRUE}
m['FB',] #1st row
```

```{r Mat31, exercise=TRUE, exercise.eval = TRUE}
m[,1] #1st column
```

```{r Mat32, exercise=TRUE, exercise.eval = TRUE}
m[,'first.names'] #1st column
```

```{r Mat33, exercise=TRUE, exercise.eval = TRUE}
m[2,1] #2nd row, 1st column
```

### Data frames
A data frame is a table where columns can be of different types but elements within a column are of a common type, like a vector. This is the type of table you'll likely use most of the time.

Can create data frame using `data.frame()`
```{r df, exercise=TRUE, exercise.eval = TRUE}
d <- data.frame(first.names, last.names, num=1:2)
d
```

```{r}
d <- data.frame(first.names, last.names, num=1:2, stringsAsFactors = TRUE)
rownames(d) <- c("FB", "CB")
```

```{r df1, exercise=TRUE, exercise.eval = TRUE}
colnames(d)
```

```{r df2, exercise=TRUE, exercise.eval = TRUE}
rownames(d) <- c("FB", "CB")
d
```

```{r df3, exercise=TRUE, exercise.eval = TRUE}
d[c("FB", "CB"), 2]
```

```{r df4, exercise=TRUE, exercise.eval = TRUE}
d[,c('first.names', 'num')]
```

We can also use `$` for data frame columns.
```{r df5, exercise=TRUE, exercise.eval = TRUE}
d$first.names
```


## Object info
Can get detailed info about objects with `summary()` or `str()`
```{r Obj, exercise=TRUE, exercise.eval = TRUE}
summary(d)
```

```{r Obj1, exercise=TRUE, exercise.eval = TRUE}
str(d) #shows structure
```

### Factors
Notice that some columns are factors. R turns characters into factors when reading/making tables by default. Factors are a special data type with levels. Factors can be difficult to use, so avoid when possible.

We can coerce factors to character.
```{r Factors, exercise=TRUE, exercise.eval = TRUE}
d[,1] <- as.character(d[,1])
str(d) #1st column is now character
```

We can also prevent R from turning characters into factors by setting a global option

```{r saf-true}
options(stringsAsFactors=TRUE)
```

```{r Factors1_5, exercise=TRUE, exercise.eval = TRUE, exercise.setup="saf-true"}
options("stringsAsFactors")
```

We change this global option to `FALSE`, which will last until the end of your R session.
```{r Factors2, exercise=TRUE, exercise.eval = TRUE}
options(stringsAsFactors=FALSE)
options("stringsAsFactors")
```

```{r saf-false}
options(stringsAsFactors=FALSE)
```

```{r Factors3, exercise=TRUE, exercise.eval = TRUE}
d2 <- data.frame(first.names, last.names, num=1:2)
str(d2) #no factors
```

## Lists
Lists are arbitrary collections of objects. The elements don't have to form a rectangular table or be of a common type. Lists can also be difficult to use, but you probably won't use them directly much.
```{r Lists, exercise=TRUE, exercise.eval = TRUE}
l <- list(pathway1=c('gene1', 'gene2'), pathway2=c('gene3', 'gene4', 'gene5'))
l
```

```{r}
l <- list(pathway1=c('gene1', 'gene2'), pathway2=c('gene3', 'gene4', 'gene5')) #Global list
```

We can select using names with `$`, like columns of a data frame (a data frame is actually a list under the hood).
```{r Lists2, exercise=TRUE, exercise.eval = TRUE}
l$pathway1
```

Some functions return lists. For example, `rownames` and `colnames` of a `matrix` are together called `dimnames`, which is a list of length 2.
```{r Lists3, exercise=TRUE, exercise.eval = TRUE}
dimnames(m)
```

```{r Lists03, exercise=TRUE, exercise.eval = TRUE}
length(dimnames(m))
```

Another function that returns a list is correlation with `cor.test()`.
```{r Lists4, eval=FALSE, exercise=TRUE}
?cor.test #in documentation, "Value" section says it returns a list
```

```{r Lists5, exercise=TRUE, exercise.eval = TRUE}
ct <- cor.test(1:3, c(6,5,7))
ct$p.value
```

## Example data set
We'll work with a simulated gene expression matrix, such as from a microarray, with 100 genes & 9 samples, split into 3 groups. We randomly simulate normally distributed values with `rnorm()`. We make the 1st gene up-regulated in group `a`.

We're going to treat this data as being on the log2-scale and already *processed*. For many datasets, there are zeroes or missing values that might need to be imputed; samples need to be normalized to make them comparable and amenable to statistical tests; absent genes need to be removed; sample outliers need to be assessed to examine whether some experimental variables or batch effects need to be accounted for or the samples need to be removed or down-weighted; and trends between a gene's mean expression and its variance should be accounted for, especially in RNA-seq data [voom](https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29).

```{r ds, exercise=TRUE, exercise.eval = TRUE}
set.seed(0) #set a seed for random sampling to get reproducible results
gx <- matrix(rnorm(n=900), nrow=100, ncol=9) #simulate data for the matrix.
gx[1:5,]
```

```{r ds1-setup}
set.seed(0) #Create a global dataset
gx <- matrix(rnorm(n=900), nrow=100, ncol=9)
```

Upregulate 1st gene in group `a`.
```{r ds1, exercise=TRUE, exercise.eval = TRUE}
gx[1, 1:3] <- gx[1, 1:3]+3

#Diplay first 5 rows of gx
gx[1:5,]
```

Add `rownames` & `colnames`
```{r ds1_5, exercise=TRUE, exercise.eval = TRUE}
#sep='' says not to put a space between 'gene' and the numbers
rownames(gx) <- paste('gene', 1:100, sep='')

#3 samples in group 'a' & 3 in group 'b' & 3 in group 'c'
colnames(gx) <- c(paste('a', 1:3, sep=''), paste('b', 1:3, sep=''), paste('c', 1:3, sep=''))
gx[1:5,]
```

```{r ds1-global}
set.seed(0) #set a seed for random sampling to get reproducible results
gx <- matrix(rnorm(n=900), nrow=100, ncol=9)
gx[1, 1:3] <- gx[1, 1:3]+3
rownames(gx) <- paste('gene', 1:100, sep='')
colnames(gx) <- c(paste('a', 1:3, sep=''), paste('b', 1:3, sep=''), paste('c', 1:3, sep=''))
gx <- data.matrix(gx)
```

We create matching phenotype & annotation data frames.
```{r ds2, exercise=TRUE, exercise.eval = TRUE}
pheno <- data.frame(grp=rep(c('a', 'b', 'c'), each=3))
pheno
```

```{r ds2-global}
pheno <- data.frame(grp=rep(c('a', 'b', 'c'), each=3)) #Global variable pheno
rownames(pheno) <- colnames(gx)
```

```{r ds21, exercise=TRUE, exercise.eval = TRUE}
rownames(pheno) <- colnames(gx)
pheno
```

```{r ds22, exercise=TRUE, exercise.eval = TRUE}
# make random gene symbols: don't worry about how I generate these
annot <- data.frame(symbol=paste0(sample(x=LETTERS, size=100, replace=TRUE), 
                                        sample(x=LETTERS, size=100, replace=TRUE), 
                                        sample(x=LETTERS, size=100, replace=TRUE)),
                    pathway=sample(x=c("oxphos", "glycolysis", "tca cycle", "none"), 
                                   size=nrow(gx), replace=TRUE))
annot
```

```{r ds22-global}
annot <- data.frame(symbol=paste0(sample(x=LETTERS, size=100, replace=TRUE), 
                                        sample(x=LETTERS, size=100, replace=TRUE), 
                                        sample(x=LETTERS, size=100, replace=TRUE)),
                    pathway=sample(x=c("oxphos", "glycolysis", "tca cycle", "none"), 
                                   size=nrow(gx), replace=TRUE))
rownames(annot) <- rownames(gx)
```

```{r ds23, exercise=TRUE, exercise.eval = TRUE}
rownames(annot) <- rownames(gx)
annot
```

## Reading / writing data
The code below writes & reads from the working directory. R assumes by default that you're reading/writing in your working directory. It works on your personal computer, but it does not work on this website.

```{r read, exercise=TRUE, exercise.eval = TRUE}
getwd() #see working directory
```

We can change working directory with `setwd()` or in RStudio under Session/Set Working Directory. Usually we write to tab-delimited text (TXT) or comma separated value (CSV). Both can be opened in Excel, and Excel can save as both formats.
```{r write, exercise=TRUE, exercise.eval = TRUE}
write.csv(gx, "gx.csv") #now look in your working directory for the file

#tables are read in by default as data frames
gx <- read.csv("gx.csv", row.names=1) #row.names=1 sets 1st column as rownames
gx[1:5,]
```

We coerce `gx` from data.frame to a numeric matrix. We could also do this with `data.matrix()`.
```{r write2, exercise=TRUE, exercise.eval = TRUE}
gx <- as.matrix(gx)
gx[1:5,]
```

## Plots
### Many types of plots are available
Can see in RStudio in bottom right *Plots* tab
```{r box, exercise=TRUE, exercise.eval = TRUE}
boxplot(gx)
```

```{r hist, exercise=TRUE, exercise.eval = TRUE}
hist(gx)
```

Most basic is `plot()`

```{r scatter, exercise=TRUE, exercise.eval = TRUE}
plot(x=gx[,1], y=gx[,2]) #scatter plot of a1 vs a2
```

We want axis labels to be sample names. We find out how to do this by looking at
```{r plot, eval=FALSE, exercise=TRUE}
?plot
```

We see that the arguments we need are:
```{r plot1, exercise=TRUE, exercise.eval = TRUE}
plot(x=gx[,1], y=gx[,2], xlab=colnames(gx)[1], ylab=colnames(gx)[2])
```

We also want to reorient axis tick labels on y-axis. 
```{r plot2, eval=FALSE, exercise=TRUE}
?plot #sends us to par. par options are applicable to many plots.
?par #there are many par options, the one we want is "las".
```

```{r plot3, exercise=TRUE, exercise.eval = TRUE}
plot(x=gx[,1], y=gx[,2], xlab=colnames(gx)[1], ylab=colnames(gx)[2], las=1)
```

### Save plot to file
We can use "Export" in *Plots* tab, or we can use code to open file, plot to it, then close it. We'll make a PDF with `pdf` function but we could use `png`, `jpeg`, `tiff`, etc.
```{r plot4, exercise=TRUE, exercise.eval = TRUE}
pdf("scatter.pdf") #open graphics device
plot(x=gx[,1], y=gx[,2], xlab=colnames(gx)[1], ylab=colnames(gx)[2], las=1)
dev.off() #close graphics device
```

### ggplot2
`ggplot2` is commonly used graphics package that makes pretty plots. You can learn more about it at [ggplot2](https://ggplot2.tidyverse.org/).
```{r ggplot, exercise=TRUE, exercise.eval = TRUE}
library(ggplot2)
gx_df <- data.frame(gx, annot)
ggp <- ggplot(data = gx_df) + geom_point(mapping=aes(x=a1, y=a2, color=pathway))
ggp
```

Save `ggplot2` plots to file with `ggsave()`.
```{r ggplot1, exercise=TRUE, exercise.eval = TRUE}
ggsave(filename = "gg_scatter.pdf", plot=ggp)
```

## Differential expression analysis
### Differential expression using base R (not add-on packages)

1. one gene at a time
2. `for` loop
3. `apply` function

### One gene at a time.
```{r exp, exercise=TRUE, exercise.eval = TRUE}
tt <- t.test(gx[1, 1:3], gx[1, 4:6]) #1st gene
str(tt)
```

Select p-value

```{r exp-pv, exercise=TRUE, exercise.eval = TRUE}
tt$p.value
```

For 2nd gene

```{r exp1, exercise=TRUE, exercise.eval = TRUE}
t.test(gx[2, 1:3], gx[2, 4:6])$p.value
```

### `for` loop
This can be automated using a `for` loop. We'll initialize a vector with `NA`s, which will store the results, & assign the t-test p-values to that vector. This loops the variable `gene` through each of the values between 1 & 100, and applies a t-test on that row.

```{r exp02, exercise=TRUE, exercise.eval = TRUE}
pv <- rep(NA, times=100)
for (gene in 1:100){
  pv[gene] <- t.test(gx[gene, 1:3], gx[gene, 4:6])$p.value
}
pv
```

### `apply`
We can also `apply` the `t.test` function to each row.
```{r exp3, exercise=TRUE, exercise.eval = TRUE}
# MARGIN=1 says to apply to rows
pv <- apply(gx, MARGIN = 1, FUN=function(row.v){ 
  t.test(row.v[1:3], row.v[4:6])$p.value 
})
pv
```

```{r exp3-global}
pv <- apply(gx, MARGIN = 1, FUN=function(row.v){ 
  t.test(row.v[1:3], row.v[4:6])$p.value 
})
```

## `ezlimma` package
You can install `ezlimma` by following instructions at [ezlimma](https://github.com/jdreyf/ezlimma). `ezlimma`'s tutorial is presented as a *vignette*, like many R packages. Here we create multiple group comparisons, and test them all in one line with `limma_contrasts`.

```{r ezlimma, exercise=TRUE}
library(ezlimma)
# browseVignettes(package="ezlimma") # would allow you to view vignette on your personal computer
```

```{r ezlimma1, exercise=TRUE, exercise.eval = TRUE}
contrast.v <- c(AvsB='a-b', AvsC='a-c', BvsC='b-c')
toptab <- limma_contrasts(object=gx, grp=pheno$grp, contrast.v = contrast.v)
toptab[1:5,]
```

```{r ezl1-global}
contrast.v <- c(AvsB='a-b', AvsC='a-c', BvsC='b-c')
toptab <- limma_contrasts(object=gx, grp=pheno$grp, contrast.v = contrast.v)
```

`toptab` is ordered by combined significance of the 3 tests. 

When we add annotation, we make sure the rows of `toptab` match the rows of `annot` with `annot[rownames(toptab),]`.

```{r ezlimma2, exercise=TRUE, exercise.eval = TRUE}
# signif() reduce the number of significant digits in toptab to 3
toptab.ann <- data.frame(signif(toptab, 3), annot[rownames(toptab),]) 
toptab.ann[1:5,]
```

```{r ezl2-global}
toptab.ann <- data.frame(signif(toptab, 3), annot[rownames(toptab),])
```

In `ezlimma`, we can also test correlation of each gene to a continuous phenotype in one line of code.

```{r ezlimma4, exercise=TRUE, exercise.eval = TRUE}
Y <- c(1:9)
toptab.y <- ezcor(object = gx, phenotype = Y, prefix = "Y")
toptab.y[1:10,]
```

`toptab.y` is ordered by the significance of the correlation and gives correlation coefficient, p-value, and FDR.

We can plot the top gene, gene27, vs. `Y`.
```{r ezlimma5, exercise=TRUE, exercise.eval = TRUE}
plot(gx["gene27",], Y)
```

For pathway analysis, `ezlimma` extends `limma` pathway analyses to write linked Excel files, as explained in the `ezlimma` vignette.

## `ezlimmaplot` package
We can plot `ezlimma` results with `ezlimmaplot`. We'll focus on the differential expression. You can install `ezlimmaplot` by following instructions at [ezlimmaplot](https://github.com/jdreyf/ezlimmaplot). `ezlimmaplot`'s tutorial is also presented as a *vignette*.

```{r elplot, exercise=TRUE}
library(ezlimmaplot)
# browseVignettes(package="ezlimmaplot")
```

### PCA
We make a Principal Component Analysis (PCA) plot to see how samples cluster. We use the `ezpca` function, which is based on the R package `ggplot2`. So that `ezpca` doesn't print to file, we set `name=NA` .

```{r elplot1, exercise=TRUE, exercise.eval = TRUE}
ezpca(object=gx, pheno.df=pheno, name=NA, color="grp")
```

Axis `PC1` captures the most variance, and `PC2` is perpendicular to `PC1` and captures the 2nd most amount of variance. This is probably the best way to see lustering in two dimensions.

### Significance histograms
We make a histogram of p-values and false discovery rates (FDR) for each comparison.
```{r elplot2, exercise=TRUE, exercise.eval = TRUE}
signif_hist(tab = toptab, name = NA)
```

We can also see what a histogram of 1,000 random p-values looks like:

```{r elplot-randopv, exercise=TRUE, exercise.eval = TRUE}
hist(x = runif(n=10^3))
```

It appears relatively flat, but not entirely.

### Venn diagram
We can see how many genes with p <= 0.05 the comparisons have in common.
```{r elplot3, exercise=TRUE, exercise.eval = TRUE}
venn.tab <- ezvenn(tab=toptab, name=NA, p.cutoff = 0.05)
```

This plot doesn't tell us which genes are in common, but `venn.tab` has this info, which we can view.

```{r elplot4, exercise=TRUE, exercise.eval = TRUE}
venn.tab[1:10,]
```

### Heatmap
We can make a heatmap of the top genes in `toptab.ann`.
```{r elplot5, exercise=TRUE, exercise.eval = TRUE}
top.genes <- rownames(toptab.ann)[1:10]
ezheat(object = gx[top.genes,], labrows=toptab.ann[1:10, "symbol"], pheno.df = pheno, name=NA)
```

```{r elplot5-global}
top.genes <- rownames(toptab.ann)[1:10]
```

Rows are labeled by gene symbol and colorbar shows groups.

By default, the row scaling *centers* the rows by subtracting the row mean. Then the new mean of every row is zero. This helps see contrast between samples, instead of between rows.

### Volcano plot
We make a volcano plot for the 1st comparison, where we label the 3 most significant genes.
```{r elplot6, exercise=TRUE, exercise.eval = TRUE}
ezvolcano(tab=toptab.ann, ntop.sig=3, lab.col="symbol", comparison = "AvsB", name=NA)
```

We can also make a volcano plot for all comparisons as a PDF with `multi_volcano(tab=toptab.ann, ntop.sig = 3, lab.col="symbol")`.

## Conclusion
Congratulations for making it this far. I hope this helped you become familiar enough with R so you can write some code, understand documentation, and hopefully analyze your data.
